src/simulator/simulator.cpp

```diff
@@ -138,50 +152,59 @@ bool Simulator::init(ros::NodeHandle& n) {
   odometryTwistMsg.header.frame_id = "odom";
   odometryTwistMsg.child_frame_id = "base_footprint";
 
-  cur_loc_ = Pose2Df(CONFIG_start_angle, {CONFIG_start_x, CONFIG_start_y});
-
-  // Create motion model based on robot type
-  // TODO extend to handle the multi-robot case
-  if (robot_type_ == "ACKERMANN_DRIVE") {
-    motion_model_ = unique_ptr<AckermannModel>(
-        new AckermannModel({CONFIG_robot_config}, &n));
-  } else if (robot_type_ == "OMNIDIRECTIONAL_DRIVE") {
-    motion_model_ = unique_ptr<OmnidirectionalModel>(
-        new OmnidirectionalModel({CONFIG_robot_config}, &n));
-  } else if (robot_type_ == "DIFF_DRIVE") {
-    motion_model_ = unique_ptr<DiffDriveModel>(
-        new DiffDriveModel({CONFIG_robot_config}, &n));
-  }
-
-  if (motion_model_ == nullptr) {
-    std::cerr << "Robot type \"" << robot_type_
-              << "\" has no associated motion model!" << std::endl;
+  if (CONFIG_robot_types.size() != CONFIG_start_poses.size()) {
+    std::cerr << "Robot type and robot start pose lists are"
+                 "not the same size!" << std::endl;
     return false;
   }
 
-  motion_model_->SetPose(cur_loc_);
+
   initSimulatorVizMarkers();
   drawMap();
 
-  initSubscriber = n.subscribe(
-      "/initialpose", 1, &Simulator::InitalLocationCallback, this);
-  odometryTwistPublisher = n.advertise<nav_msgs::Odometry>("/odom",1);
-  laserPublisher = n.advertise<sensor_msgs::LaserScan>(CONFIG_laser_topic, 1);
-  viz_laser_publisher_ = n.advertise<sensor_msgs::LaserScan>("/scan", 1);
-  mapLinesPublisher = n.advertise<visualization_msgs::Marker>(
-      "/simulator_visualization", 6);
-  posMarkerPublisher = n.advertise<visualization_msgs::Marker>(
-      "/simulator_visualization", 6);
-  objectLinesPublisher = n.advertise<visualization_msgs::Marker>(
-      "/simulator_visualization", 6);
-  truePosePublisher = n.advertise<geometry_msgs::PoseStamped>(
-      "/simulator_true_pose", 1);
-  if (FLAGS_localize) {
-    localizationPublisher = n.advertise<ut_multirobot_sim::Localization2DMsg>(
-        "/localization", 1);
-    localizationMsg.header.frame_id = "map";
-    localizationMsg.header.seq = 0;
+  // Create motion model based on robot type
+  // TODO extend to handle the multi-robot case
+  for (size_t i = 0; i < CONFIG_start_poses.size(); ++i) {
+    const auto& robot_type = CONFIG_robot_types.at(i);
+    const auto& start_pose = CONFIG_start_poses.at(i);
+    const auto pf = IndexToPrefix(i);
+    auto* mm = MakeMotionModel(robot_type, n, pf);
+    if (mm == nullptr) {
+      return false;
+    }
+    mm->SetPose(Pose2Df(start_pose.z(), {start_pose.x(), start_pose.y()}));
+
+    robot_pub_subs_.emplace_back(RobotPubSub());
+    auto& rps = robot_pub_subs_.back();
+    rps.motion_model = std::unique_ptr<robot_model::RobotModel>(mm);
+
+    rps.initSubscriber = n.subscribe<PoseWithCovarianceStamped>(
+       pf + "/initialpose", 1, [&](const boost::shared_ptr<const PoseWithCovarianceStamped>& msg) {
```

Comment:
Why do we need boost::shared_ptr?

Reply:
This is the type that is specified concretely for lambdas; when you do regular function pointers it's marshaling it into a different type. I tried marshaling it into a different type such as a raw pointer or a const reference but got compiler errors.